test/common/network/dns_impl_test.cc

```diff
@@ -107,31 +124,78 @@ class TestDnsServerQuery {
         const std::list<std::string>* ips = nullptr;
         // We only expect resources of type A or AAAA.
         const int q_type = DNS_QUESTION_TYPE(question + name_len);
+        std::string cname;
+        // check if we have a cname
+        auto cit = parent_.cnames_.find(name);
+        if (cit != parent_.cnames_.end()) {
+          cname = cit->second;
+        }
+        const char* hostLookup = name;
+        const unsigned char* ip_question = question;
+        long ip_name_len = name_len;
+        std::string encodedCname;
+        if (cname.size() > 0) {
+          hostLookup = cname.c_str();
+          encodedCname = TestDnsServerQuery::encodeDnsName(cname);
+          ip_question = reinterpret_cast<const unsigned char*>(encodedCname.c_str());
+          ip_name_len = encodedCname.size() + 1;
+        }
         ASSERT_TRUE(q_type == T_A || q_type == T_AAAA);
         if (q_type == T_A) {
-          auto it = parent_.hosts_A_.find(name);
+          auto it = parent_.hosts_A_.find(hostLookup);
           if (it != parent_.hosts_A_.end()) {
             ips = &it->second;
           }
         } else {
-          auto it = parent_.hosts_AAAA_.find(name);
+          auto it = parent_.hosts_AAAA_.find(hostLookup);
           if (it != parent_.hosts_AAAA_.end()) {
             ips = &it->second;
           }
         }
         ares_free_string(name);
 
+        int answer_size = ips != nullptr ? ips->size() : 0;
+        answer_size += encodedCname.size() > 0 ? 1 : 0;
+
         // The response begins with the intial part of the request
         // (including the question section).
         const size_t response_base_len = HFIXEDSZ + name_len + QFIXEDSZ;
         unsigned char response_base[response_base_len];
         memcpy(response_base, request, response_base_len);
         DNS_HEADER_SET_QR(response_base, 1);
         DNS_HEADER_SET_AA(response_base, 0);
-        DNS_HEADER_SET_RCODE(response_base, ips != nullptr ? NOERROR : NXDOMAIN);
-        DNS_HEADER_SET_ANCOUNT(response_base, ips != nullptr ? ips->size() : 0);
+        DNS_HEADER_SET_RCODE(response_base, answer_size > 0 ? NOERROR : NXDOMAIN);
+        DNS_HEADER_SET_ANCOUNT(response_base, answer_size);
         DNS_HEADER_SET_NSCOUNT(response_base, 0);
         DNS_HEADER_SET_ARCOUNT(response_base, 0);
+        size_t response_ip_rest_len;
+        if (q_type == T_A) {
+          response_ip_rest_len =
+              ips != nullptr ? ips->size() * (ip_name_len + RRFIXEDSZ + sizeof(in_addr)) : 0;
+        } else {
+          response_ip_rest_len =
+              ips != nullptr ? ips->size() * (ip_name_len + RRFIXEDSZ + sizeof(in6_addr)) : 0;
+        }
+        size_t response_cname_len =
+            encodedCname.size() > 0 ? name_len + RRFIXEDSZ + encodedCname.size() + 1 : 0;
+        // Send response to client.
```

Comment:
This comment makes this section a little confusing. 

Previously, we computed all the elements of the response and then wrote them all to the client. 

In this PR we have moved to computing and writing in lockstep on each piece of the request. 

I think we should either refactor to how it was previously done (compute then write all),  or change the comments (the one this review comment hangs off of, add a comment above line 171 `size_t response_ip_rest_len`, etc)

Reply:
I agree the cname flow is a bit confusing. First there is need to see if a domain name resolves to a cname entry. Then there is need to lookup the cname record to see if it resolves to an ip or multiple ips. We still need to compute the total response size before starting to write each response element. This is computed with the variable: `response_size_n`. After writing out the header response, then we need to write out all response elements which include the cname entry and the ips. I'll add a comment to better explain this logic.